### PR TITLE
feat(clients/go): add metrics for job worker

### LINKED ISSUE #4500: Metrics about scheduled jobs
**Is your feature request related to a problem? Please describe.**
We want to provide metrics about our zeebe workers to better monitor them. Specifically, how many jobs a worker has scheduled (`jobPoller`).

**Describe the solution you'd like**
The `JobWorker` offers a possibility to register a consumer implementation of an interface predefined by the zeebe go client

**Describe alternatives you've considered**
IMHO there is no alternative because the information about remaining jobs is currently not exposed

### PR DESCRIPTION
**Pull request by AndreasBieber**
## Description

Introduces an interface `JobWorkerMetrics` which can be registered with the `JobWorkerBuilder`:
```
builder.Metrics(metricsImpl)
```

## Related issues

closes #4500

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [ ] ~~If submitting code, please run `mvn clean install -DskipTests` locally before committing~~ (no java code changes)

### CODE REVIEW

File: clients/go/pkg/worker/jobPoller_test.go [hunk 1]
```diff
@@ -204,6 +204,30 @@ func (suite *JobPollerSuite) TestShouldIgnoreStreamFailure() {
 	suite.consumeJob()
 }
 
+func (suite *JobPollerSuite) TestShouldReportMetrics() {
```
Comment (npepinpe) [inline clients/go/pkg/worker/jobPoller_test.go:207]: Is this testing that the metric is reset to 0 on close?

**Review by npepinpe — CHANGES_REQUESTED** [review]
Thanks a lot for this PR and your patience :slightly_smiling_face:  

LGTM, can you also add a test for the new builder method in `jobWorker_builder_test.go`?

One small question, I'll admit I'm not a Go expert - as you can see most of our project is in Java. This is the first time we add metrics in the Go client; on the Java side, I'd usually recommend using something like Micrometer as a metrics facade - is there something similar in the Go ecosystem?

File: clients/go/pkg/worker/jobPoller_test.go (hunk 1 above)
Comment (AndreasBieber) [reply clients/go/pkg/worker/jobPoller_test.go:207]: Yes, the test includes also resetting the metric to 0 on close.

**Comment by AndreasBieber** [comment]
We usually use the Prometheus client to expose metrics, as it is the de facto standard in the cloud native environment (`github.com/prometheus/client_golang`).

But I did not want to introduce a new dependency in the zeebe go client.

**Comment by AndreasBieber** [comment]
Actually I wanted to pass a metric facade to the client. However, this would have resulted in breaking changes, because the abstraction is not sufficiently separated. This would have had the advantage to capture much more metrics. Perhaps this can be addressed in the future.

**Comment by npepinpe** [comment]
> We usually use the Prometheus client to expose metrics, as this is the de facto standard in the cloud native environment (github.com/prometheus/client_golang).
> 
> But I did not want to introduce a new dependency in the zeebe go client.

Totally agree - we made that (imo) mistake in our main project but I would love to replace that with Micrometer.

> Actually I wanted to pass a metric facade to the client. However, this would have resulted in breaking changes, because the abstraction is not sufficiently separated. This would have had the advantage to capture much more metrics. Perhaps this can be addressed in the future.

I'd be happy to discuss what you have in mind there :slightly_smiling_face:

**Comment by npepinpe** [comment]
Tested it with Prometheus, looks good :+1:

**Review by npepinpe — APPROVED** [review]

**Review by npepinpe — CHANGES_REQUESTED** [review]
One small thing, sorry I didn't flag it before; we keep track of backwards compatibility, and this breaks it (you can read more [here](https://github.com/zeebe-io/zeebe/wiki/Breaking-Changes#go)). Now that's partially because we're Java devs and using interfaces in Go in the wrong way (for which we have an issue, see [here](https://github.com/zeebe-io/zeebe/issues/3741#issue-555677508)).

In my opinion, this change is worth breaking, as the benefits outweigh the gains, and I doubt many users are implementing the job worker builder interface.

If you could then run 

```sh
gocompat save ./...
```

And push that; you can verify it now passes the compat check by running

```sh
gocompat compare --go1compat ./...
```

Thanks! Sorry for the back and forth :sweat_smile:

**Review by npepinpe — APPROVED** [review]
Thanks so much, and thanks for being patient with us :slightly_smiling_face: 

This will be part of the next release, 0.24, scheduled for the first week of July, though there weren't much changes since 0.23 in the Go client so you could probably use (at your own risk) the commit tag.

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
Sorry, pesky flaky tests :sweat_smile: 

bors retry

**Pull request merged**